src/library/rdt/src/rdt_specialsxp.cpp

```diff
@@ -0,0 +1,117 @@
+#include <stdlib.h>
+#include <map>
+#include <fstream>
+
+extern "C" {
+// If I don't include this before rdt.h, I get strange compiler errors...
+#include "r.h"
+#include "rdt.h"
+}
+
+#include "rdt_register_hook.h"
+
+#include "rdt_promises/tracer_conf.h"
+#include "rdt_promises/tracer_output.h"
+
+static std::map<std::string, uint64_t> specialsxp_count;
+
+struct trace_specialsxp {
+
+    DECL_HOOK(begin)(const SEXP prom) {
+    }
+
+    DECL_HOOK(end)() {
+        ofstream myfile;
+        myfile.open ("specialsxp_analysis.txt");
+        for (auto &pair : specialsxp_count) {
+            Rprintf("%s : %llu\n", pair.first.c_str(), pair.second);
+            myfile << pair.first << " : " << pair.second << "\n";
+        }
+        myfile.close();
+    }
+
+    DECL_HOOK(function_entry)(const SEXP call, const SEXP op, const SEXP rho) {
+    }
+
+    DECL_HOOK(function_exit)(const SEXP call, const SEXP op, const SEXP rho, const SEXP retval) {
+    }
+
+    DECL_HOOK(builtin_entry)(const SEXP call, const SEXP op, const SEXP rho) {
+    }
+
+    DECL_HOOK(builtin_exit)(const SEXP call, const SEXP op, const SEXP rho, const SEXP retval) {
+    }
+
+    DECL_HOOK(specialsxp_entry)(const SEXP call, const SEXP op, const SEXP rho) {
+
+        std::string call_name = std::string(get_name(call));
+
+        if (specialsxp_count.find(call_name) == specialsxp_count.end()) {
+            specialsxp_count[call_name] = 1;
+        } else {
+            specialsxp_count[call_name]++;
+        }
+    }
+
+    DECL_HOOK(specialsxp_exit)(const SEXP call, const SEXP op, const SEXP rho, const SEXP retval) {
+    }
+
+    DECL_HOOK(force_promise_entry)(const SEXP symbol, const SEXP rho) {
+    }
+
+    DECL_HOOK(force_promise_exit)(const SEXP symbol, const SEXP rho, const SEXP val) {
+    }
+
+    DECL_HOOK(promise_lookup)(const SEXP symbol, const SEXP rho, const SEXP val) {
+    }
+
+    DECL_HOOK(error)(const SEXP call, const char* message) {
+    }
+
+    DECL_HOOK(vector_alloc)(int sexptype, long length, long bytes, const char* srcref) {
+    }
+
+    DECL_HOOK(gc_entry)(R_size_t size_needed) {
+    }
+
+    DECL_HOOK(gc_exit)(int gc_count, double vcells, double ncells) {
+    }
+
+    DECL_HOOK(S3_generic_entry)(const char *generic, const SEXP object) {
+    }
+
+    DECL_HOOK(S3_generic_exit)(const char *generic, const SEXP object, const SEXP retval) {
+    }
+
+    DECL_HOOK(S3_dispatch_entry)(const char *generic, const char *clazz, const SEXP method, const SEXP object) {
+    }
+
+    DECL_HOOK(S3_dispatch_exit)(const char *generic, const char *clazz, const SEXP method, const SEXP object, const SEXP retval) {
+    }
+};
+
+
+rdt_handler *setup_specialsxp_tracing(SEXP options) {
+    rdt_handler *h = (rdt_handler *)  malloc(sizeof(rdt_handler));
+    *h = REGISTER_HOOKS(trace_specialsxp,
```

Comment:
No, actually you don't have to declare all of them. We can remove the empty functions.

I even changed the hook registration so that you can also enumerate only the hooks you use instead of initializing the whole structure icluding the `NULL` slots. That was annoying, because everytime you added a hook, you had to update all the existing tracers by adding more `NULL`s.